### ISSUE TITLE
undefined method 'read' for DataMapper::Mash

### ISSUE THREAD

**Issue by bunnymatic** [krobertson/dm-paperclip#23]
Not sure if this is exactly a paperclip issue, but using code that was working to upload to local filesystem, i just moved to use S3 and can't seem to get anything working.  Here's the error i get

```
 NoMethodError - undefined method `read' for #<DataMapper::Mash:0x000001031e7c48>:
/Users/jon/.rvm/gems/ruby-1.9.2-p290@romey/bundler/gems/dm-paperclip-d92cc6cd0ac5/lib/dm-paperclip/attachment.rb:250:in `generate_fingerprint'
/Users/jon/.rvm/gems/ruby-1.9.2-p290@romey/bundler/gems/dm-paperclip-d92cc6cd0ac5/lib/dm-paperclip/attachment.rb:117:in `assign'
/Users/jon/.rvm/gems/ruby-1.9.2-p290@romey/bundler/gems/dm-paperclip-d92cc6cd0ac5/lib/dm-paperclip.rb:334:in `block in has_attached_file'
```

This is pulling from your latest (the gem file refers directly to the git repo).

The mash that's fed in there is constructed from the file upload form like so

```
  mash = DataMapper::Mash.new
  mash['tempfile'] = file_hash[:tempfile]
  mash['filename'] = file_hash[:filename]
  mash['content_type'] = file_hash[:type]
  mash['size'] = file_hash[:tempfile].size
```

where the file_hash is the params[:file] from the upload form.

Any ideas on what might be going on here?

Other specs:  Sinatra, ruby 1.9.2, datamapper 1.2.0

Cheers
Jon

**Comment by krobertson** [comment]
Fixed in aabad04f2a65e8f628e7e838d9bd3e1c9c39914d and pushed to master.  Confirmed there was a hole in test coverage with hash/mash like objects and it was a bug with calling generate_fingerprint twice and the first one not being compatible.

**Comment by bunnymatic** [comment]
awesome!

good customer service.  I was going to dig in last night, but i was
just a little to tired.

thanks for the quick fix.

jon

On Thu, May 17, 2012 at 9:43 AM, Ken Robertson
reply@reply.github.com
wrote:

> Fixed in aabad04f2a65e8f628e7e838d9bd3e1c9c39914d and pushed to master.  Confirmed there was a hole in test coverage with hash/mash like objects and it was a bug with calling generate_fingerprint twice and the first one not being compatible.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/krobertson/dm-paperclip/issues/23#issuecomment-5768136

## 

Visit http://bunnymatic.com
